Source/OnlineSubsystemEpic/Private/OnlineFriendsInterfaceEpic.h

```diff
@@ -0,0 +1,232 @@
+ *
+ * NOTE: these functions only support accounts on EGS or those
+ * linked to an EGS account with a different supported identity provider.
+ */
+class FOnlineFriendEpic : public FOnlineFriend
+{
+
+public:
+	/**
+	 * Init/default constructor that must take in an Epic Account Id
+	 */
+	FOnlineFriendEpic(const EOS_EpicAccountId& InUserId = EOS_EpicAccountId());
+
+	FOnlineFriendEpic(const FUniqueNetId& NetId);
+
+	/** Virtual destructor to keep resharper happy */
+	virtual ~FOnlineFriendEpic() {};
+
+	// FOnlineUser
+	virtual TSharedRef<const FUniqueNetId> GetUserId() const override;
+	virtual FString GetRealName() const override;
+	virtual FString GetDisplayName(const FString& Platform = FString()) const override;
+
+	/**
+	 * Get account data attribute
+	 *
+	 * @param Key account data entry key
+	 * @param OutVal [out] value that was found
+	 *
+	 * @return true if entry was found
+	 */
+	virtual bool GetUserAttribute(const FString& AttrName, FString& OutAttrValue) const override;
+
+	// FOnlineFriend
+
+	virtual EInviteStatus::Type GetInviteStatus() const override;
+	virtual const FOnlineUserPresence& GetPresence() const override;
+
+	// Custom functions
+	EFriendStatus GetFriendStatus() const;
+
+	virtual void FORCEINLINE SetFriendStatus(EFriendStatus InFriendStatus) { FriendStatus = InFriendStatus; }
+	virtual void FORCEINLINE SetPresence(FOnlineUserPresence InPresence) { Presence = InPresence; }
+
+	bool SetUserLocalAttribute(const FString& AttrName, const FString& InAttrValue) override;
+
+protected:
+	/** User Id represented as a FUniqueNetId */
+	TSharedRef<const FUniqueNetId> UserId;
+	
+	/** Any additional account data associated with the friend */
+	TMap<FString, FString> AccountData;
+	
+	/** Friend status that can be used for blueprints */
+	EFriendStatus FriendStatus;
+	
+	/* Full presence information including platform type, application, status, rich text. */
+	FOnlineUserPresence Presence;
+};
+
+class FOnlineFriendsInterfaceEpic : public IOnlineFriends
+{
+
+public:
+	FOnlineFriendsInterfaceEpic() : Subsystem(nullptr),
+	                               friendsHandle(nullptr)
+	{
+	}
+
+	/**
+	 * Initializes the various interfaces
+	 *
+	 * @param InSubsystem the subsystem that owns this object
+	 */
+	FOnlineFriendsInterfaceEpic(FOnlineSubsystemEpic* InSubsystem);
+
+	virtual ~FOnlineFriendsInterfaceEpic() {};
+
+	/** The subsystem that owns the instance */
+	FOnlineSubsystemEpic* Subsystem;
+	
+	/** Delegates for non-friend users that were called with QueryPresence */
+	TMap<FUniqueNetIdEpic, IOnlinePresence::FOnPresenceTaskCompleteDelegate> DelayedPresenceDelegates;
+
+protected:
+	EOS_HFriends friendsHandle;
+	FOnReadFriendsListComplete ReadFriendsListDelegate;
+
+	/** List of friends */
+	struct FEpicFriendsList
+	{
+		TArray<TSharedRef<FOnlineFriendEpic> > Friends;
+	};
+	
+	/**Mmap of local user idx to friends */
+	TMap<int32, FEpicFriendsList> FriendsLists;
```

Comment:
Use [TMultiMap](https://docs.unrealengine.com/en-US/API/Runtime/Core/Containers/TMultiMap/index.html) instead.

Reply:
Why? SteamFriendsSubsystem does it like this and I don't see any problems with it. I only need one type - the struct is there for readability.

Reply:
With `TMultiMap` You can write: `TMultiMap<int32, FEpicFriendsList> FriendsLists;` and then use methods like `MultiFind` to get the values for a key. It does the same thing you do, but without the need for an extra struct or nested `TArray`

Reply:
Hm, maybe I can add it later as an enhancement. Would like to see what other issues people experience or want added. For instance, one of the enhancements I was considering adding was a way to just pass in "Online" or "Offline" and then filter out all of those friends that have offline or online. This is something that could only be done in EAS since no other service provides this, but I didn't deem it necessary for now. I could definitely use a MultiMap for then